### PR TITLE
Throws SecurityTokenMalformedTokenException on malformed tokens

### PR DESCRIPTION
**Pull request by dannybtsai**
Currently an ArgumentException is thrown when a token handler fails to parse/read a raw token. Since the ArgumentException is a general exception for any invalid arguments, sometimes it does not provide specific types of errors, ex, invalid header or payload. This PR adds a new exception, SecurityTokenMalformedTokenException, that can be thrown when a malformed token is detected. For example, the exception will be thrown if a SAML token is handled by JsonWebTokenHandler.

This should provide a clearer exception so users to handle it differently.

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/Exceptions/SecurityTokenMalformedTokenException.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using System;
+using System.Runtime.Serialization;
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Represents a <see cref="SecurityToken"/> exception when the token is malformed.
+    /// </summary>
+    [Serializable]
+    public class SecurityTokenMalformedTokenException : SecurityTokenArgumentException
```
Comment (victorm-hernandez) [inline src/Microsoft.IdentityModel.Tokens/Exceptions/SecurityTokenMalformedTokenException.cs]: >SecurityTokenMalformedTokenException [](http://example.com/codeflow?start=17&length=36)

I think the name is a bit redundant including the word token twice.

Lets rename it to something like SecurityTokenMalformedException? #Resolved

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs [hunk 2]
```diff
@@ -427,17 +427,14 @@ private void ReadToken(string encodedJson)
...
+            if (Dot1 == -1 || Dot1 == encodedJson.Length - 1)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedTokenException(LogHelper.FormatInvariant(LogMessages.IDX14100, encodedJson)));
 
             Dot2 = encodedJson.IndexOf('.', Dot1 + 1);
             if (Dot2 == -1)
-                throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14120, encodedJson)));
+                throw LogHelper.LogExceptionMessage(new SecurityTokenMalformedTokenException(LogHelper.FormatInvariant(LogMessages.IDX14120, encodedJson)));
 
-            if (Dot2 == encodedJson.Length)
+            if (Dot2 == encodedJson.Length - 1)
```
Comment (victorm-hernandez) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:444]: >if (Dot2 == encodedJson.Length - 1) [](http://example.com/codeflow?start=12&length=35)

This change is unrelated with the exceptions thrown, are we fixing a bug here? #Resolved

**Comment by victorm-hernandez** [comment]
>         /// <exception cref="ArgumentException">'jwtEncodedString' contains only whitespace.</exception>

Let's update the XML docs to reflect we are throwing a different exception type.

---
In reply to: [1551947809](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551947809) [](http://example.com/codeflow?ancestors=1551947809)

---
In reply to: [1551947809](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551947809) [](http://example.com/codeflow?ancestors=1551947809)

---
Refers to: src/System.IdentityModel.Tokens.Jwt/JwtSecurityToken.cs:24 in 1f748de. [](commit_id = 1f748dea213e2cac880c34cdfae827dccafb9929, deletion_comment = False)

**Comment by victorm-hernandez** [comment]
>         /// <exception cref="ArgumentException"><paramref name="token"/>.Length is greater than <see cref="TokenHandler.MaximumTokenSizeInBytes"/>.</exception>

ditto: lets update XML docs for the new exception type

---
In reply to: [1551949110](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551949110) [](http://example.com/codeflow?ancestors=1551949110)

---
In reply to: [1551949110](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551949110) [](http://example.com/codeflow?ancestors=1551949110)

---
Refers to: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:815 in 1f748de. [](commit_id = 1f748dea213e2cac880c34cdfae827dccafb9929, deletion_comment = False)

**Comment by victorm-hernandez** [comment]
>         public override ClaimsPrincipal ValidateToken(string token, TokenValidationParameters validationParameters, out SecurityToken validatedToken)

Let's update also the JsonWebTokenHandler and Saml2SecurityTokenHandler ValidateToken methods to be consistent.

---
In reply to: [1551955852](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551955852) [](http://example.com/codeflow?ancestors=1551955852)

---
Refers to: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:835 in 1f748de. [](commit_id = 1f748dea213e2cac880c34cdfae827dccafb9929, deletion_comment = False)

**Review by victorm-hernandez — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs (hunk 2 above)
Comment (dannybtsai) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:444]: Yes, this is a bug that should be fixed.

Comment (victorm-hernandez) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:444]: Sounds good, lets make sure the commit message also reflects we are doing this fix.

Comment (dannybtsai) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebToken.cs:444]: I already committed and updated the branch. I think I can update the comment when it is ready to be merged.

**Comment by victorm-hernandez** [comment]
>         public override ClaimsPrincipal ValidateToken(string token, TokenValidationParameters validationParameters, out SecurityToken validatedToken)

Danny created a work item for this
[https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2589877](https://identitydivision.visualstudio.com/Engineering/_workitems/edit/2589877)

---
In reply to: [1551955852](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2080#issuecomment-1551955852) [](http://example.com/codeflow?ancestors=1551955852,1551955852)

---
Refers to: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:835 in 1f748de. [](commit_id = 1f748dea213e2cac880c34cdfae827dccafb9929, deletion_comment = False)

**Review by victorm-hernandez — APPROVED** [review]
:shipit:
